### PR TITLE
Fix Knative/OpenShift client environment check

### LINKED ISSUE #28171: KnativeDeployer environment check is breaking the build
### Describe the bug

We have apps that use Knative client and now with `kubernetes-client` dependency has been migrated to 6.1.1, the builds are broken. It might be down to changes descried in migration guide https://github.com/fabric8io/kubernetes-client/blob/master/doc/MIGRATION-v6.md#adapt-changes

### Expected behavior

Builds just fine.

### Actual behavior

Exception is thrown:
```
...
12:56:59,665 INFO  mvn: Caused by: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
12:56:59,665 INFO  mvn: 	[error]: Build step io.quarkus.kubernetes.deployment.KnativeDeployer#checkEnvironment threw an exception: java.lang.IllegalStateException: Knative was requested as a deployment, but the target cluster is not a Knative cluster!
12:56:59,665 INFO  mvn: 	at io.quarkus.kubernetes.deployment.KnativeDeployer.lambda$checkEnvironment$0(KnativeDeployer.java:29)
12:56:59,665 INFO  mvn: 	at java.base/java.util.Optional.ifPresent(Optional.java:178)
12:56:59,665 INFO  mvn: 	at io.quarkus.kubernetes.deployment.KnativeDeployer.checkEnvironment(KnativeDeployer.java:21)
12:56:59,665 INFO  mvn: 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
12:56:59,665 INFO  mvn: 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
12:56:59,666 INFO  mvn: 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
12:56:59,666 INFO  mvn: 	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
12:56:59,666 INFO  mvn: 	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:909)
12:56:59,666 INFO  mvn: 	at io.quarkus.builder.BuildContext.run(BuildContext.java:281)
12:56:59,666 INFO  mvn: 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
12:56:59,666 INFO  mvn: 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
12:56:59,666 INFO  mvn: 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
12:56:59,666 INFO  mvn: 	at java.base/java.lang.Thread.run(Thread.java:833)
12:56:59,666 INFO  mvn: 	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
12:56:59,667 INFO  mvn: 
12:56:59,667 INFO  mvn:     at io.quarkus.builder.Execution.run (Execution.java:123)
12:56:59,667 INFO  mvn:     at io.quarkus.builder.BuildExecutionBuilder.execute (BuildExecutionBuilder.java:79)
...
```

### How to Reproduce?

Reproducer:

0. Have OC stuff ready
1. `git clone https://github.com/quarkus-qe/quarkus-test-suite.git`
2. `cd http/http-minimum`
3. `mvn clean verify -Dit.test=ServerlessExtensionOpenShiftHttpMinimumIT -Dopenshift -Dserverless -Doperator-scenarios`

I can provide standalone reproducer if required.

### Output of `uname -a` or `ver`

Linux fedora 5.19.6-200.fc36.x86_64 #1 SMP PREEMPT_DYNAMIC Wed Aug 31 17:58:15 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "17.0.3" 2022-04-19

### GraalVM version (if different from Java)

GraalVM CE 22.1.0

### Quarkus version or git rev

999-SNAPSHOT

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.6

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by michalvavrik**
fixes: #28171

According to https://github.com/fabric8io/kubernetes-client/blob/master/doc/MIGRATION-v6.md#adapt-changes, `isAdaptable` method semantic has changed (and the method is deprecated) and it's causing failures in Quarkus QE Test Suite. AFAIK originally `isAdaptable` check wasn't strict (`contains`, `endsWith`), so you could have ApiGroups like (see below) and check would succeed https://github.com/fabric8io/kubernetes-client/commit/bb6115a75a6d42df19d1f99953791c4dd26eb472#diff-a23a9372cb2d1a1c4121272460b2c8e2f8dd79bf5722b8b2a36ba5d4578cef75 . Now, check fails if there is no group that strictly matches `APIGroup` `knative.dev`. That's however not necessary for our app to work, f.e. our APIGroups are:
- serving.knative.dev
- serving.knative.dev/v1
- serving.knative.dev/v1
- serving.knative.dev/v1beta1
- serving.knative.dev/v1alpha1
- autoscaling.internal.knative.dev
- autoscaling.internal.knative.dev/v1alpha1
- autoscaling.internal.knative.dev/v1alpha1
- caching.internal.knative.dev
- caching.internal.knative.dev/v1alpha1
- caching.internal.knative.dev/v1alpha1
- networking.internal.knative.dev
- networking.internal.knative.dev/v1alpha1
- networking.internal.knative.dev/v1alpha1
- operator.knative.dev
- operator.knative.dev/v1alpha1
- operator.knative.dev/v1alpha1

When Quarkus is not performing strict tests for `knative.dev`, our app is working and check performed is same as before.

### CODE REVIEW

**Review by Sgitario — COMMENTED** [review]
Many many thanks for your pull request!

Spites I understand this is broken and that there might not have a better solution, I think these changes are a bit fragile (I mean checking the coordinates of the fabric8:openshift-client artifact). 

About the problem itself, I don't fully follow what exactly changed in regards to the `isAdaptable` method. @manusa can you help here?

Also, I'm not sure if this is possible, but adding some tests would help. 

Anyway, let's wait for @manusa's feedback.

**Comment by michalvavrik** [comment]
@Sgitario I'll try to write a test than (I've already tried, but sure, there must be a way...). As for `isAdaptable`, all I know is what changed https://github.com/fabric8io/kubernetes-client/commit/bb6115a75a6d42df19d1f99953791c4dd26eb472#diff-d3ca8c6252f8f76bcd09c8faa8b0039c49886b929de4388ed0ca05df521f01c1L37 and it does cause troubles. And this PR is what I understood a migration guide suggest. It's very much possible I misunderstood it! Let's here from expert.

**Comment by michalvavrik** [comment]
My understanding is that `isAdaptable` is now calling OC, which is surplus to requirements when just checking if we can adapt client to KnativeClient (ex. from test I wrote, not from reproducer):
```
...
Caused by: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.kubernetes.deployment.KnativeDeployer#checkEnvironment threw an exception: io.fabric8.kubernetes.client.KubernetesClientException: An error has occurred.
	at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:129)
	at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:122)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.restCall(OperationSupport.java:779)
	at io.fabric8.kubernetes.client.BaseClient.getApiGroup(BaseClient.java:242)
	at io.fabric8.kubernetes.client.BaseClient.hasApiGroup(BaseClient.java:158)
	at io.fabric8.kubernetes.client.extension.ClientAdapter.hasApiGroup(ClientAdapter.java:73)
	at io.fabric8.knative.client.DefaultKnativeClient.isSupported(DefaultKnativeClient.java:244)
	at io.fabric8.kubernetes.client.BaseClient.isAdaptable(BaseClient.java:176)
	at io.quarkus.kubernetes.deployment.KnativeDeployer.lambda$checkEnvironment$0(KnativeDeployer.java:26)
	at java.base/java.util.Optional.ifPresent(Optional.java:178)
	at io.quarkus.kubernetes.deployment.KnativeDeployer.checkEnvironment(KnativeDeployer.java:21)

...
```

**Comment by michalvavrik** [comment]
@Sgitario I changed PR and its description. This fix should require tests (that if you think about it are hard to do). I'd suggest to wait for response to @manusa and eventually close this PR if it's wrong/someone else provides better one.

**Comment by manusa** [comment]
Couldn't read the code changes yet, but a quick note on the `isAdaptable`, and in general Adapt, behavioral changes:

> **Adapt Changes**
> `Client.isAdaptable` and `Client.adapt` will check first if the existing client is an instance of the desired type.
> `Client.adapt` will no longer perform the `isAdaptable` check - that is you may freely adapt from one Client to another as long as the extension exists. If you need to make a specific check of support, please use the `Client.supports` method.


You can read more about them at our [Migration Guide](https://github.com/fabric8io/kubernetes-client/blob/master/doc/MIGRATION-v6.md).

This means that the behavior for this method has changed completely for version 6.

The `isAdaptable` method is deprecated and should no longer be used.

If you need to check if the cluster supports the operations that the derived client supports (in this case I guess `KnativeClient` or `OpenShiftClient`), you should use the `.isSupported()` method instead. (`kubernetesClient.adapt(KnativeClient.class).isSupported()`)

In addition, you can use the `.supports(Class<R> type)` with any of our generated model classes, or the `.hasApiGroup(String apiGroup, boolean exact)` method.

File: extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KnativeDeployer.java
Comment (michalvavrik) [reply extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KnativeDeployer.java:25]: But it's strict, previously the check wasn't so strict (please see PR description). With Quarkus 2.13.0.CR1, our app worked. I saw the logic behind `ExtensionAdapterSupport` and it wasn't strict (endsWith, contains). Am I wrong? @manusa

Comment (michalvavrik) [reply extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KnativeDeployer.java:25]: That's why I went with `false` (and it also fixes reproducer :-))

Comment (manusa) [reply extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KnativeDeployer.java:25]: Yup, you're right.
I think we might need to fix this in the client extension too.

Comment (manusa) [reply extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KnativeDeployer.java:25]: https://github.com/fabric8io/kubernetes-client/issues/4447

**Review by manusa — APPROVED** [review]

File: extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KnativeDeployer.java
Comment (michalvavrik) [reply extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KnativeDeployer.java:25]: thank you

**Comment by michalvavrik** [comment]
@Sgitario I want to urge you in no way, just give me a hint what's next (or ETA) as if you need a time, I'm going to disable QE tests, thanks a lot

**Comment by manusa** [comment]
Just a quick note that today is a bank holiday in Barcelona

**Comment by michalvavrik** [comment]
Ah, thank you @manusa

File: extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KnativeDeployer.java
Comment (Sgitario) [reply extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KnativeDeployer.java:25]: @michalvavrik please add a comment with the link of the pull request, so we can change it in the future.

File: extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/OpenshiftDeployer.java
Comment (Sgitario) [inline extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/OpenshiftDeployer.java:36]: I don't think we encourage users to use non-Quarkus dependencies to make extensions work. 
When could this exception be thrown? Maybe when users use the `quarkus-kubernetes` extension and select `openshift` as the deployment target? If so, better to let users use the `quarkus-openshift` extension. Moreover, the sentence "Failed to adapt `OpenShiftClient`" is a bit technical, better to write a comprehensible and general message: "Failed to configure OpenShift. Make sure you have the Quarkus OpenShift extension.".

**Review by Sgitario — COMMENTED** [review]
Apart from the comments, the changes look good to me. 
Many thanks @michalvavrik !

File: extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/OpenshiftDeployer.java
Comment (michalvavrik) [reply extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/OpenshiftDeployer.java]: Sure, that makes much more sense. Done.

File: extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KnativeDeployer.java
Comment (michalvavrik) [reply extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KnativeDeployer.java:26]: I added a comment with the link to the issue, hope that's okay (once issue is closed, we will know PR was merged...). Thanks

**Comment by michalvavrik** [comment]
Thank you for the review.

**Review by Sgitario — APPROVED** [review]

**Comment by Sgitario** [comment]
Many thanks @michalvavrik . lgtm now
@geoand are you ok with merging this one?

**Review by geoand — APPROVED** [review]
Seems like you folks have figured it out.

Feel free to merge when CI is green

**Comment by Sgitario** [comment]
Failures are unrelated to this PR. Merging.

**Comment by gsmet** [comment]
@Sgitario @geoand @michalvavrik I think this should be backported, right?

**Comment by michalvavrik** [comment]
This behavior comes from switch to Kubernetes client 6.1.1, thus 2.13 should not be affected.

**Comment by manusa** [comment]
Note that these methods will need refactoring when bumping the client to 6.5 or at most 6.6 

https://github.com/fabric8io/kubernetes-client/pull/4869

**Comment by geoand** [comment]
Thanks for the heads up!

**Pull request merged**
